cep-20.2.md

```diff
@@ -0,0 +1,832 @@
+strong: [PackageSelector]
+# weak run exports go from host -> run
+weak: [PackageSelector]
+# strong constraints adds a run constraint from build -> run_constraints (was `strong_constrains`)
+strong_constraints: [PackageSelector]
+# weak constraints adds a run constraint from host -> run_constraints (was `weak_constrains`)
+weak_constraints: [PackageSelector]
+# noarch run exports go from host -> run for `noarch` builds
+noarch: [PackageSelector]
+```
+
+### Test section
+
+<details>
+<summary>
+The current state of the Test section which is being removed with this spec.
+
+Note: the test section also has a weird implicit behavior with `run_test.sh`, `run_test.bat` as well as `run_test.py` and `run_test.pl` script files
+that are run as part of the test.
+
+</summary>
+The current YAML format for the test section is:
+
+```yaml
+test:
+  # files (from recipe directory) to include with the tests
+  files: [glob]
+  # files (from the work directory) to include with the tests
+  source_files: [glob]
+  # requirements at test time, in the target_platform architecture
+  requires: [PackageSelector]
+  # commands to execute
+  commands: [string]
+  # imports to execute with python (e.g. `import <string>`)
+  imports: [string]
+  # downstream packages that should be tested against this package
+  downstreams: [PackageSelector]
+```
+
+</details>
+
+The new test section consists of a list of test elements. Each element is executed independently and can have different requirements.
+There are multiple types of test elements defined, such as the `command` test element, the `python` test element and the `downstream` test element.
+
+Before, the test section was written to a single folder (`info/test`). In the new format, we propose to write each test element to a separate folder
+(`info/tests/<index>`). This allows us to run each test element independently.
+
+```yaml
+tests: [TestElement]
+```
+
+#### Command test element
+
+The command test element renders to a single folder with a `test_time_dependencies.json` with two keys (`build` and `run`) that contain the raw "PackageSelector" strings.
+The `script` is rendered to a `script.json` that contains the `interpreter`, `env` and other keys (as defined in the `Script` section).
+Files are copied into the `info/tests/<index>` folder.
+
+```yaml
+# script to execute
+# reuse script definition from above
+script: string | [string] | Script
+# optional extra requirements
+requirements:
+  # extra requirements with build_platform architecture (emulators, ...)
+  build: [PackageSelector]
+  # extra run dependencies
+  run: [PackageSelector]
+
+# extra files to add to the package for the test
+files:
+  # files from $SRC_DIR
+  source: [glob]
+  # files from $RECIPE_DIR
+  recipe: [glob]
+```
+
+#### Python test element
+
+The python test element renders a `test_import.py` file that contains the imports to test.
+It also automatically runs the `pip check` command to check for missing dependencies.
```

Comment:
This means that `pip` must be always injected in the test environment? Or is it running from the tool (rattler-build) environment? This might cause hard-to-debug conflicts in the test env if the user's package is not compatible with part of the `pip` dependency tree (arguably a very small fraction, but I thought I should bring it up).

Reply:
I'm also not keen on automatically running `pip check`, especially since optional features and dependencies are a thing. Need to dig out the specific package name, but we have run into issues where `pip check` failed because we (as packagers) explicitly decided to turn off a feature but had no way of informing pip of that change.